### ISSUE TITLE
Using `Client.share.rx` functions in versions above 2.0.1 not working

### ISSUE THREAD

**Issue by Cordavi** [GetStream/stream-chat-swift#245]
## What did you do?

- I updated from `StreamChat` `2.0.1` to `2.1.1` and no longer can make calls like `Client.shared.rx.queryChannels(filter: .currentUserInMembers).subscribe`.

## What did you expect to happen?

- I would expect the call to still call when a channel changes. I also expect that making any call on `Client.shared.rx` does not cause my `ChannelsViewController` with a presenter of `ChannelsPresenter(filter: .currentUserInMembers)` to stop working and showing user's chats.
 
## What happened instead?

- When I make calls on `Client.shared.rx` like querying the channels, this function no longer calls and I don't get subscription updates. Also, making this call is causing me to lose all conversations in my `ChannelsViewController`. 

## GetStream Environment
**GetStream Chat version:** 2.0.1
**iOS version:** 13.4.1
**Swift version:** 5
**Xcode version:** 11.4.1
**Device:** Simulator / iPhone 11 Pro

**Comment by b-onc** [comment]
Hello @Cordavi ,

In my attempt to reproduce the issue, I've added this code:
```swift
Client.shared.rx.queryChannels(filter: .currentUserInMembers).take(1).subscribe { (event) in
            print(event)
}.disposed(by: disposeBag)
```
to run when a button is pressed in `ChannelsViewController`. The call completes successfully each time and and I don't lose the channels.

Can you provide more info, steps to reproduce, code samples for this issue?

**Comment by Cordavi** [comment]
Hi @b-onc,

I attempt to run this call after the stream user is set but before a ‘ChannelsViewController’ has been created. 

First controller calls this in view did load, after a call to setup the stream user. Then a user taps a button to present a ChannelsViewController. 

Hope that helps.

**Comment by buh** [comment]
First of all, you have to set a user before doing any query or subscriptions.
Please don't do `disconnect()` (it's like user logout) until you need to re-login a user.
If you do disconnect you have to re-subscribe for updates.
Can you tell me more about your login/logout, queries/subscriptions flow?

To subscribe for the global unread count for the user subscribe for this (as in Example app):
```swift
        Client.shared.rx.unreadCount
            .observeOn(MainScheduler.instance)
            .subscribe(
                onNext: { [weak self] unreadCount in
                    // Total unread channels \(unreadCount.channels), messages: \(unreadCount.messages)"
                },
                onError: { [weak self] in ... })
            .disposed(by: disposeBag)

```

**Comment by buh** [comment]
Hi @Cordavi 
I made a PR #249 
Can you check the branch?

**Comment by Cordavi** [comment]
Hi  @buh,

I tested out the branch and it did not resolve my issue.

I did however find a potential workaround by adjusting the timing of my code to make sure we do not add a subscriber before `Client.shared.isConnected == true`. If I wait for that to be `true` then I get the calls and I start seeing my `ChannelsViewController`’s `channel`’s for the user again.

It seems like something changed in the framework from `2.0.1` to `2.1.0+` that is causing the timing of when I’m setting up the stream user, when it becomes connected, and when I'm setting up this subscriber to monitor the channel's to no longer work.

**Comment by buh** [comment]
Yes, the flow changed in 2.1
You have to subscribe after set user. As I said above: https://github.com/GetStream/stream-chat-swift/issues/245#issuecomment-624624271
Can you check that?

**Comment by Cordavi** [comment]
The flow I have goes roughly like this:

1. We get the current user. On completion of getting the current user we setup the subscriber.
```swift
        userManager.getCurrentUser { [weak self] result in
            switch result {
            case .success:
                Client.shared.rx.queryChannels(filter: .currentUserInMembers).subscribe { event in
                    DispatchQueue.main.async { [weak self] in
                        self?.channelsWithoutMessagesCount = event.element?.filter { $0.messages.isEmpty }.count ?? 0
                    }
                }
            case .failure:
                DispatchQueue.main.async { [weak self] in
                    self?.channelsWithoutMessagesCount = 0
                }
            }
        }
```

2. As a part of getting the current user we set the stream user.
```swift
    private func configureChatUser() {
        let userExtraData = UserExtraData(name: user?.name)
        let user = StreamChatClient.User(id: uid, extraData: userExtraData)
        let tokenProvider: TokenProvider = { [weak self] supplyToken in
            self?.requestJWT { result in
                switch result {
                case .success(let token):
                    supplyToken(token)
                case .failure:
                    break
                }
            }
        }
        Client.shared.set(user: user, tokenProvider: tokenProvider)
    }
```

3. The `Client.shared.set(user: user, tokenProvider: tokenProvider)` call happens before the completion where I setup the subscriber. However, I find that `isConnected` is not `true` by the time the completion calls.
 - Note that we can call `Client.shared.set(user: user, tokenProvider: tokenProvider)` multiple times. Is there side effects of calling this multiple times?

**Comment by buh** [comment]
Hi @Cordavi 
We released 2.2.0 with fix #249.

1. The completion block from `userManager.getCurrentUser` should be called after `Client.shared.set(user: user, tokenProvider: tokenProvider)`, but you no need to wait for the successful connection. Basically, when your user logged out you should cancel all subscriptions and setup them again when your new user logged in. In practice, that means your UI shouldn't be shown (subscribed) for not logged in user. 

2. A small tip for your code:
```swift
var streamUser = StreamChatClient.User(id: uid)
streamUser.name = user?.name
```
3. You shouldn't call `Client.shared.set(user: user, tokenProvider: tokenProvider)` multiple times for the same user. Only when you need to relogin a user. For user logout only, you need to call `Client.shared.disconnect()`.

Let me know if you have more questions.

**Comment by b-onc** [comment]
Closed via #284
